daemon/src/rollover/taker.rs

```
@@ -0,0 +1,223 @@
+///
+/// If the maker does not respond within `DECISION_TIMEOUT` seconds then the taker will fail the
+/// rollover.
+const DECISION_TIMEOUT: Duration = Duration::from_secs(30);
+
+/// One actor to rule all the rollovers
+pub struct Actor {
+    endpoint: Address<Endpoint>,
+    oracle_pk: schnorrsig::PublicKey,
+    get_announcement: Box<dyn MessageChannel<oracle::GetAnnouncement>>,
+    n_payouts: usize,
+    tasks: Tasks,
+    executor: command::Executor,
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[derive(Copy, Clone)]
+pub struct ProposeRollover {
+    pub order_id: OrderId,
+    pub maker_peer_id: PeerId,
+}
+
+impl Actor {
+    pub fn new(
+        endpoint: Address<Endpoint>,
+        executor: command::Executor,
+        oracle_pk: schnorrsig::PublicKey,
+        get_announcement: Box<dyn MessageChannel<oracle::GetAnnouncement>>,
+        n_payouts: usize,
+    ) -> Self {
+        Self {
+            endpoint,
+            tasks: Tasks::default(),
+            executor,
+            get_announcement,
+            oracle_pk,
+            n_payouts,
+        }
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    pub async fn handle(&mut self, msg: ProposeRollover) {
+        let ProposeRollover {
+            order_id,
+            maker_peer_id,
+        } = msg;
+
+        let substream = match self
+            .endpoint
+            .send(OpenSubstream::single_protocol(
+                maker_peer_id.inner(),
+                rollover::PROTOCOL,
+            ))
+            .await
+            .context("Endpoint is disconnected")
+            .context("Failed to open substream")
+        {
+            Ok(Ok(substream)) => substream,
+            // TODO: Same error path and add failed
+            Ok(Err(e)) => {
+                tracing::error!("Failed to open connection to maker: {e:#}");
+                return;
+            }
```

🔧 I think it will be more convenient to extract this into an `Actor` method, use `?` and handle all errors with a single `if let Err(e)` in this `xtra` handler.

---

Given that we are on the taker we could consider it. Generally I would like to avoid the pattern of sending messages to ourselves because it adds unnecessary overhead.

---

I extracted a little helper (I understood @luckysori comment as extracting a helper for a bit nicer error handling). It make things a bit more clear, as it flattened the match statement and made the code more readable by not repeating the code.
I opted for not excluding the whole handler as we still need to `emit_fail` on the error - if it was just logging, I'd return an error from the handler and put everything in the method.